UKBiobank ingestion

hlnicholls opened the pull request:
I have worked on the UKBiobank ingestion, making these changes:

- Added a `StudyIndexUKBiobank` class in `study_index.py`
- Created `ukbiobank.py` containing a `UKBiobankStep` class
- Added a `UKBiobankStepConfig` class in `config.py`
- Created a `my_ukbiobank.yaml` for configuration
- Added the UKBiobank inputs and outputs for the ingestion in `gcp.yaml`
- Included a 'my_ukbiobank' id in `dag.yaml`
- Added UKBiobank study index and configuration tests (in `test_study_index.py` and `conftest.py`)

I have also updated the `README.md` to add in a line about needing java version 8 on one-time configuration.

I've tested the ingestion code both locally and in the workflow. Pytest also seems to run with no problems. The output schema is:

```
root
 |-- studyId: string (nullable = true)
 |-- traitFromSource: string (nullable = true)
 |-- nCases: long (nullable = true)
 |-- initialSampleSize: string (nullable = true)
 |-- summarystatsLocation: string (nullable = true)
 |-- publicationDate: string (nullable = false)
 |-- pubmedId: string (nullable = false)
 |-- publicationJournal: string (nullable = false)
 |-- publicationTitle: string (nullable = false)
 |-- publicationFirstAuthor: string (nullable = false)
 |-- discoverySamples: array (nullable = false)
 |    |-- element: struct (containsNull = false)
 |    |    |-- sampleSize: string (nullable = true)
 |    |    |-- ancestry: string (nullable = true)
 |-- replicationSamples: array (nullable = false)
 |    |-- element: struct (containsNull = false)
 |    |    |-- sampleSize: string (nullable = false)
 |    |    |-- ancestry: string (nullable = false)
 |-- projectId: string (nullable = false)
 |-- studyType: string (nullable = false)
 |-- hasSumstats: boolean (nullable = false)
 ```
An example of the first output 10 rows looks like:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/hn9/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/hn9/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:1.0in .75in 1.0in .75in;
	mso-header-margin:.5in;
	mso-footer-margin:.5in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{mso-number-format:"Short Date";}
-->

</head>

<body link="#0563C1" vlink="#954F72">



studyId | traitFromSource | nCases | initialSampleSize | summarystatsLocation | publicationDate | pubmedId | publicationJournal | publicationTitle | publicationFirstAuthor | discoverySamples | replicationSamples | projectId | studyType | hasSumstats
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
NEALE2_100001_raw | Food weight |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100001_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100002_raw | Energy |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100002_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100003_raw | Protein |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100003_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100004_raw | Fat |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100004_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100005_raw | Carbohydrate |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100005_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100006_raw | Saturated fat |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100006_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100007_raw | Polyunsaturated fat |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100007_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100008_raw | Total sugars |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100008_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100009_raw | Englyst dietary fibre |   | 51453 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100009_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51453', ancestry='European=51453')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE
NEALE2_100010 | Portion size |   | 51427 | gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100010.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz | 01/08/2018 |   |   |   | UKB Neale v2 | [Row(sampleSize='51427', ancestry='European=51427')] | [Row(sampleSize='', ancestry='')] | UKBiobank | gwas | TRUE



</body>

</html>

d0choa commented:
Beautiful! The failing test looks like you forgot to commit the `data_samples/neale2_saige_study_manifest.samples.tsv` file.

Can you try to add the `docs` sections for the `step` and the `dataset`? It will take you 2 min.
You can locally inspect the result using the following:
`poetry run mkdocs serve`

tskir commented on src/otg/dataset/study_index.py:
```diff
@@ -477,3 +477,113 @@ def from_source(
+        """
+        return cls(
+            _df=(
+                ukbiobank_studies.select(
+                    f.col("code").alias("studyId"),
+                    f.col("trait").alias("traitFromSource"),
+                    f.col("n_cases").cast("long").alias("nCases"),
+                    f.col("n_total").cast("string").alias("initialSampleSize"),
+                    f.col("in_path").alias("summarystatsLocation"),
+                )
+                # Replace any trait double spaces with single.
+                .withColumn(
+                    "traitFromSource",
+                    f.regexp_replace(f.col("traitFromSource"), r" +", " "),
+                )
```
As @ireneisdoomed and @d0choa once explained to me, it's usually a good idea to replace multiple `.withColumn` statements with a bigger `select` statment. So in this case, for example, I would change it like this:

```suggestion
                ukbiobank_studies.select(
                    f.col("code").alias("studyId"),
                    # Replace any trait double spaces with single.
                    f.regexp_replace(f.col("trait"), r" +", " ").alias("traitFromSource"),
                    f.col("n_cases").cast("long").alias("nCases"),
                    f.col("n_total").cast("string").alias("initialSampleSize"),
                    f.col("in_path").alias("summarystatsLocation"),
                )
```

The reason is that each `withColumn` statments adds a bit of computational complexity, whereas a big `select` is more efficient

tskir replied:
This would also apply to some (but not all) of the `withColumn` statements below. Essentially, if there's an easy way to move it to the `select` statement, do it; but if it seems tricky or even not possible (which it could be, in some cases), just leave them be

tskir commented on src/otg/dataset/study_index.py:
```diff
@@ -477,3 +477,113 @@ def from_source(
+                        "2018-08-01",
+                    ).otherwise("2018-10-24"),
+                )
+                .withColumn("pubmedId", f.lit(""))
+                .withColumn("publicationJournal", f.lit(""))
+                .withColumn("publicationTitle", f.lit(""))
```
I assume the reason we're leaving those fields blank is that the UK Biobank studies don't have an associated literature information?

@d0choa Can you please remind me what do we want to do in this case? Just leave them blank? Or perhaps hardcode one citation for the entire UK Biobank?

(Although I just discovered that they don't have a paper on that for some reason, and are just asking to cite... the URL: http://www.nealelab.is/uk-biobank/faq)

ireneisdoomed replied:
NEALEs' certainly don't have a publication, but we can referece SAIGE with [30104761](https://pubmed.ncbi.nlm.nih.gov/30104761/)

Instead of empty strings, I would leave these fields as null.

hlnicholls replied:
Thank you for this, just to be sure, should I add in the SAIGE reference and leave all other fields as null, something like:
```
.withColumn("pubmedId",
                    f.when(
                        f.col("studyId").startswith("SAIGE_"),
                        "30104761",
                    ).otherwise(None),
                )

.withColumn("publicationJournal",
                    f.when(
                        f.col("studyId").startswith("SAIGE_"),
                        "Nature Genetics",
                    ).otherwise(None),
                )

.withColumn("publicationTitle",
                    f.when(
                        f.col("studyId").startswith("SAIGE_"),
                        "Efficiently controlling for case-control imbalance and sample relatedness in large-scale genetic association studies",
                    ).otherwise(None),
                )
```
Or instead just leave all as null with:
```
.withColumn("pubmedId", f.lit(None))
.withColumn("publicationJournal", f.lit(None))
.withColumn("publicationTitle", f.lit(None))
```

tskir commented on src/otg/dataset/study_index.py:
```diff
@@ -477,3 +477,113 @@ def from_source(
+                    f.col("trait").alias("traitFromSource"),
+                    f.col("n_cases").cast("long").alias("nCases"),
+                    f.col("n_total").cast("string").alias("initialSampleSize"),
+                    f.col("in_path").alias("summarystatsLocation"),
```
When I try to access the sumstats from the gs:// links generated by this part, I'm getting:

```
$ gsutil ls gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/100001_raw.neale2.gwas.imputed_v3.both_sexes.GRCh38.tsv.split*.gz
CommandException: One or more URLs matched no objects.
```

(This URL is from the first row in the PR description)

tskir replied:
So in this case, I wonder if they have an updated manifest file somewhere?

tskir commented on tests/data_samples/neale2_saige_study_manifest.samples.tsv:
```diff
@@ -0,0 +1,11 @@
+in_path	code	trait	n_total	n_cases	trait_category
+gs://genetics-portal-raw/uk_biobank_sumstats/harmonised_split_190318/PheCode_244_SAIGE_Nov2017_MACge20.GRCh38.tsv.split*.gz	SAIGE_244	Hypothyroidism	406300	14871	endocrine/metabolic
```
Yep, also `CommandException: One or more URLs matched no objects.` for these ones.

@d0choa @DSuveges Do you know if this might be due to the fact that I just don't have access? Or is it likely that they moved them somewhere?

(I _think_ sumstats are supposed to be public anyway, right?)

tskir reviewed (commented):
Great PR all around! Just added some notes on Spark optimisation, citations, and summary stats location

ireneisdoomed commented on src/otg/dataset/study_index.py:
```diff
@@ -477,3 +477,113 @@ def from_source(
+                # Replace any trait double spaces with single.
+                .withColumn(
+                    "traitFromSource",
+                    f.regexp_replace(f.col("traitFromSource"), r" +", " "),
```
I'm not sure what this function does. As I understand, it removes any extra whitespace, however the data looks the same to me after applying it. Am I understanding things correctly?

hlnicholls replied:
It should be removing whitespace, I was trying to bring over the same function from the old make_UKB_study_table.py [here](https://github.com/opentargets/genetics-v2d-data/blob/4b5e6f9989dca73f93dcdeaa8fc8d1bd3f51690f/scripts/make_UKB_study_table.py#L90) - but I can go ahead and remove this if it's not needed

ireneisdoomed replied:
I don't think it is, tbf

ireneisdoomed commented on src/otg/dataset/study_index.py:
```diff
@@ -477,3 +477,113 @@ def from_source(
+                    ),
+                )
+                .withColumn(
+                    "replicationSamples",
```
I'd remove this whole statement. If we lack information for `replicationSamples`, just avoid the field while it is nullable

ireneisdoomed commented on src/otg/ukbiobank.py:
```diff
@@ -0,0 +1,31 @@
+"""Step to run ukb study table ingestion."""
```
```suggestion
"""Step to run UKBiobank study table ingestion."""
```

ireneisdoomed reviewed (changes requested):
Awesome! All the boxes are ticked. As mentioned in the meeting, the only thing burning is the wrong sumstats locations, but we can tackle that in another PR. I also have questions about the provenance of that manifest, if we produce it and we are going to change it anyways, we might as well put it in a format that is more aligned to our schema now.

In any case, I'd leave that to a diff PR. I have some minor comments regarding null handling, other than that it is good to go! Thank you!

hlnicholls commented on README.md:
```diff
@@ -18,6 +18,8 @@ Google Cloud configuration:
 
 Check that you have the `make` utility installed, and if not (which is unlikely), install it using your system package manager.
 
+Ensure that you have `java` version 8 installed on your system (also known as java 1.8).
```
Oh maybe it doesn't need to just be version 8. I was having an issue running pytest with Java 20. @tskir and I talked about it and downgrading to version 8 solved my error (which was `Py4JJavaError(
py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.`).

The PySpark documentation says 'PySpark requires Java version 1.8. 0 or the above version' so not sure why I had an issue with version 20. Maybe we can reject this change in the README for now? I can look more into if there is a specific requirement or if I just had something else java-related going wrong.

tskir replied:
I think what's happening here is that while PySpark declares it needs version 8 (and it indeed works on 8), it actually unofficially supports some higher versions as well. (Anecdotally, I encountered some other examples of software that require JRE 8 on paper but actually run fine on 10-11.) But evidently changes in version 20 are too large and it can't support that anymore.

I would definitely include some change to the README mentioning Java, because when someone has an incorrect version, it doesn't just say "Wrong Java version", it generates a _really_ long and cryptic traceback which is quite hard to make sense of.

I would revise the change to say something like: "PySpark requires Java version 8, a. k. a. 1.8, to run propertly. It may also work on higher versions up to at least 11, but will not work on high versions like Java 20. If you are encountering problems with initialising a Spark session, make sure you have the correct version of Java, ideally Java 8."

tskir replied:
Actually, on re-reading PySpark requirements, it doesn't look like it's version locked. So it probably will work on most versions, and the issue with 20 is that it's _too_ new (released in March 23), and Spark didn't have time to adapt. So then the revised revised statement would read:

> PySpark requires Java version 8 (a.k.a. 1.8) or above to work. However, it may not work with the most recent Java versions, like Java 20 at the time of writing this section (May 2023). If you are encountering problems with initialising a Spark session, try using Java 11 or Java 8.

hlnicholls commented on config/datasets/gcp.yaml:
```diff
@@ -20,6 +20,7 @@ catalog_studies: ${datasets.inputs}/v2d/gwas-catalog-v1.0.3-studies-r2022-11-29.
 catalog_ancestries: ${datasets.inputs}/v2d/gwas-catalog-v1.0.3-ancestries-r2022-11-29.tsv
 catalog_sumstats_lut: ${datasets.inputs}/v2d/harmonised_list.txt
 finngen_phenotype_table_url: https://r8.finngen.fi/api/phenos
+ukbiobank_manifest: gs://genetics-portal-input/ukb_phenotypes/neale2_saige_study_manifest.190430.tsv
```
I'm gonna double check this with the genetics team. From googling the file name in quotes the only search result is the v2d repo so I suspect we generated it - but I will let you know what I find out

hlnicholls commented:
I think I have now addressed most of the comments, thank you @tskir and @ireneisdoomed for your feedback! Here is what has been updated in my `fix: updated StudyIndexUKBiobank` commit (alongside adding all the commit suggestions):
- Updated the docstring with the UKBB Neale and SAIGE study details
- Removed the `replicationSamples` column creation
- Removed casting as string and setting `otherwise(None)`
- Removed the whitespace removal function
- Updated the publication columns to include SAIGE information

I've also updated the java version recommendation in `docs/contributing.md`

The only thing I haven't been able to do is to reduce the amount of `withColumn()` for `select()`. I have some local code that runs with only `select()` (getting rid of all the `withColumn()`) and it runs without problems but on submitting to the workflow gives an error I haven't been able to solve yet - so maybe something to work on in another PR to get this one merged?

tskir reviewed (approved):
As all of the comments have now been addressed, I'm approving and merging this. Thank you Hannah, exemplary work
